ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -320,6 +327,92 @@ private static Object instantiateUdfClass(final Method method,
     }
   }
 
+  private static Object instantiateUdfClass(final Class udfClass,
+                                            final UdfDescription annotation) {
+    try {
+      return udfClass.newInstance();
+    } catch (final Exception e) {
+      throw new KsqlException("Failed to create instance for UDF="
+              + annotation.name(), e);
+    }
+  }
+
+  private Function<List<Schema>,Schema> handleUdfReturnSchema(
+          final Class theClass,
+          final Method method,
+          final Udf udfAnnotation,
+          final UdfDescription descAnnotation) {
+
+    final String schemaProviderName = method.getAnnotation(Udf.class).schemaProvider();
+
+    if (!schemaProviderName.equals("")) {
+      return handleUdfSchemaProviderAnnotation(schemaProviderName, theClass, method,
+              descAnnotation);
+    }
+
+    final Schema returnSchema = getReturnType(method, udfAnnotation);
+    if (returnSchema.name() != null
+        && returnSchema.name().equals("org.apache.kafka.connect.data.Decimal")
+        && schemaProviderName.equals("")) {
+      throw new KsqlFunctionException("BigDecimal return type is not supported "
+          + "without a schema provider method.");
+    }
+
+    return ignored -> returnSchema;
+
+  }
+
+  private Function<List<Schema>,Schema> handleUdfSchemaProviderAnnotation(
+          final String schemaProviderName,
+          final Class theClass,
+          final Method method,
+          final UdfDescription annotation) {
+
+    Function<List<Schema>, Schema> schemaProviderFunction = null;
+
+    final List<Method> allMethods = new ArrayList<Method>(
+            Arrays.asList(theClass.getDeclaredMethods()));
+
+    // throws exception if cannot find method
+    final Method m = findSchemaProvider(allMethods, schemaProviderName);
+    final Object instance = instantiateUdfClass(theClass, annotation);
+
+    schemaProviderFunction = list -> {
```

Comment:
```suggestion
    schemaProviderFunction = parameterTypes -> {
```
nit: should have meaningful names - especially in lambdas - if possible